python/cudf/tests/test_reductions.py

```diff
@@ -0,0 +1,149 @@
+from __future__ import division, print_function
+
+import pytest
+import random
+import numpy as np
+
+from itertools import product
+from cudf.dataframe import Series
+from cudf.tests import utils
+from cudf.tests.utils import gen_rand
+
+
+params_dtype = [
+    np.float64,
+    np.float32,
+    np.int64,
+    np.int32,
+    np.int16,
+    np.int8,
+]
+
+params_sizes = [1, 2, 3, 127, 128, 129, 200, 10000]
+
+params = list(product(params_dtype, params_sizes))
+
+
+@pytest.mark.parametrize('dtype,nelem', params)
+def test_sum(dtype, nelem):
+    data = gen_rand(dtype, nelem)
+    sr = Series(data)
+
+    got = sr.sum()
+    expect = dtype(data.sum())
+
+    print('expect:', expect)
+    print('got:', got)
+
+    significant = 4 if dtype == np.float32 else 6
+    np.testing.assert_approx_equal(expect, got, significant=significant)
+
+
+@pytest.mark.parametrize('dtype,nelem', params)
+def test_product(dtype, nelem):
+    if np.dtype(dtype).kind == 'i':
+        data = np.ones(nelem, dtype=dtype)
+        # Set at most 30 items to [0..2) to keep the value within 2^32
+        for _ in range(30):
+            data[random.randrange(nelem)] = random.random() * 2
+    else:
+        data = gen_rand(dtype, nelem)
+
+    sr = Series(data)
+
+    got = sr.product()
+    expect = np.product(data)
+
+    print('expect:', expect)
+    print('got:', got)
+
+    significant = 4 if dtype == np.float32 else 6
+    np.testing.assert_approx_equal(expect, got, significant=significant)
+
+
+accuracy_for_dtype = {
+    np.float64: 6,
+    np.float32: 5
+}
+
+
+@pytest.mark.parametrize('dtype,nelem', params)
+def test_sum_of_squares(dtype, nelem):
+    data = gen_rand(dtype, nelem)
+    sr = Series(data)
+
+    got = sr.sum_of_squares()
+    expect = (data ** 2).sum()
+
+    print('expect:', expect)
+    print('got:', got)
+
+    if np.dtype(dtype).kind == 'i':
+        if 0 <= expect <= np.iinfo(dtype).max:
+            np.testing.assert_array_almost_equal(expect, got)
+        else:
+            print('overflow, passing')
```

Comment:
If we overflow the specified dtype what happens here? Does it silently overflow? Should we have some type of typecasting rules?

Reply:
Silently overflows.
It is documented.
Using higher output precision is recommended if possible.
https://github.com/rapidsai/cudf/blob/d5c7105946244a5171bb5b36357f338077183bfb/cpp/include/reduction.hpp#L46-L69

Reply:
#1418